### ISSUE TITLE
Duplicate task list when syncing todo.txt using Syncthing

### ISSUE THREAD

**Issue by dekonst** [ransome1/sleek#125]
When syncing todo.txt using Syncthing, sleek displays duplicate task list. When updating todo.txt locally using text editor, everything is correct.

**To Reproduce**
1. Create sync folder using Syncthing
2. Place todo.txt inside sync folder
3. Add Task 1 using sleek (Screenshot 1)
4. Add Task 2 using sleek on another computer
5. Wait for todo.txt to be synced
6. See error on Screenshot 2

**Do you see any error entries in sleeks developer tools?**
No errors.

**Expected behavior**
Correct display of updated todo.txt.

**Screenshots**
![Task 1](https://user-images.githubusercontent.com/33258381/117683169-79c73b00-b1bc-11eb-8a1a-334ac59ad93c.png)
![Task 2](https://user-images.githubusercontent.com/33258381/117683287-982d3680-b1bc-11eb-9e01-6967c8f8e4fc.png)

**Desktop (please complete the following information):**
 - OS: Manjaro
 - Version of sleek 1.0.3
 - Source: Github, AppImage
 - Syncthing: v1.15.1, Linux (64-bit Intel/AMD)

**Comment by ransome1** [comment]
And another question: When this happens and you refresh your todo list, will the duplicated items disappear?

**Comment by dekonst** [comment]
Unfortunately, I didn't find how to refresh todo list... Please, let me know how to do that. Test.txt file I've sent to your address.

**Comment by ransome1** [comment]
@dekonst, you can refresh by pressing Ctrl and R.

The good news is, I was able to reproduce it and I'm working on a solution.

**Comment by ransome1** [comment]
Hi @dekonst. Could you please take a look at this testing release, if it solves the issue on your environment? https://github.com/ransome1/sleek/releases/tag/v1.0.3-1

**Comment by dekonst** [comment]
@ransome1, I've just checked Syncthing logs, I don't have any warnings.

But my environment is slightly different - another computer is my mobile phone with Android and Syncthing app, and I change the todo.txt file on it using text editor.

On my PC I have Double Commander, it signals me when the file is changed. I have sleek and text editor opened over it. After changes on my phone it takes about 7 second to sync the file. When I see that modification time of the file is changed in Double Commander, I reload file in text editor manually and see the changes in it. But after second change as I told nothing happens in sleek.

I really don't see any difference in local and remote changes. The only guess is timings and file locks. I could imagine that if Syncthing opens file in exclusive mode, changes file content and then releases file lock after some time, then sleek could be notified by file system that file is changed and it will try to read its content. In this case if file lock is not released yet, sleek could get file access error that could put sleek into some state where it doesn't monitor file changes until manual refresh.

But it's only the guess... Of course, its better to try to reproduce this behavior using the same environment. But if it for some reason couldn't be done, I can try to check it with debug version of sleek if it can write trace log.

I even tried to build sleek from source using your instructions but haven't succeeded in it.

**Comment by ransome1** [comment]
@dekonst, you're right I need to reproduce it, because I don't think the logs will produce any error.

What I understand from your setup is, that you basically do this:
1. On Computer 1 you create the 1st todo -> this todo will be saved locally and synced successfully to Computer 2
2. After 1.: On Computer 2 you create the second todo -> This todo will be saved locally but won't be synced to Computer 2. Plus it breaks the whole function of saving new todos (add, edit, mark as complete, etc.) on Computer 2 as well

This is how I understand the issue on your side, correct?

One thing you can try is to open up the Developer Console (Press Alt, then the top menu will appear and in the tab "About" you select "Developer Tools"), then break the function. So if for instance you then mark a todo as complete, which should not work, I wonder if there is an error displayed in the console tab of the Dev Tools.

Also I assume the process you see on this video doesn't work for you?

https://user-images.githubusercontent.com/11188741/117922874-6bc80600-b2f3-11eb-99a2-dd8dff1b6ca5.mp4

(my local Arch Linux on the left and a Manjaro VM on the right)

**Comment by dekonst** [comment]
Hi @ransome1! I'm sorry if I was not clear. Let me describe what I'm doing. I'm trying to check how sleek monitors file changes.

Scenario 1:
1. Start sleek from command line in terminal.
2. Open todo.txt in text editor on the same computer.
3. Change and save todo.txt, it could be just adding one space at the end of the file.
4. For every file save I see exactly one Info line in the terminal.

![LocalChanges](https://user-images.githubusercontent.com/33258381/118043378-f5241a80-b37d-11eb-930f-dac322055f5d.png)

Scenario 2:
1. Start sleek from command line in terminal.
2. Open todo.txt in text editor on Android device.
3. Change and save todo.txt.
4. Wait for the file to be synced.
5. Two Info lines appear in the terminal at once for the first sync.
6. Nothing happens in the terminal and in Developer Console for the following file changes.
7. Filewatcher doesn't detect any following changes of the file even made locally.
8. It looks like Filewatcher suspends after that.
9. Filewatcher starts working only after it is restarted by sleek refresh with Ctrl + R (See second entry with "Filewatcher is watching" string).

![SyncthingChanges](https://user-images.githubusercontent.com/33258381/118045585-c6f40a00-b380-11eb-8b98-1cc5461ebd31.png)


It looks like two sequential file changes made by Syncthing somehow break Filewatcher. But I haven't succeeded yet to prepare the repro case without Syncthing.

As a dirty workaround I could propose to restart Filewatcher after every detected file change, if it's possible, but I guess that it could be not good idea...

**Comment by dekonst** [comment]
@ransome1, good news, I've got an answer.

The problem is that you monitor changes to the todo.txt file itself.

What I have if I monitor file changes with inotify-tools utility:

![FileWatch](https://user-images.githubusercontent.com/33258381/118059791-07ac4d00-b39a-11eb-8c62-dbb1872c83ee.png)

After that no changes to the file is detected.

If I monitor directory changes, then the answer is clear:

![DirectoryWatch](https://user-images.githubusercontent.com/33258381/118059895-39bdaf00-b39a-11eb-822e-1843b85cb746.png)

When Syncthing detects remote changes, it first writes new content to the temporary file, then deletes old file and after that renames temporary file into original file name. In this case after delete operation old file doesn't exist and no changes to the new file are detected.

So, the solution would be to monitor changes to the directory where todo.txt file is located and filter related events by file name in that directory.

This would be more complicated to implement but reliable to detect file changes.

**Comment by ransome1** [comment]
@dekonst. Thanks for the analysis on this. But I have my doubts if just changing the watch method (with all the implementation effort behind it) will solve this issue. Plus I can't reproduce it exactly the way you did. I use Simplenote in combination with Syncthing on my Android device. And in scenario 2, after the changes in Android are being synced to my computer, the filewatcher continues working on my side. I can still work with sleek and local and remote changes are being recognized.

Your tool states "DELETE SELF". Do you understand what exactly happens here? Is syncthing actually deleting (and apparantly breaking the filewatcher) your old file shortly before it's being replaced? Because when I log the actual JavaScript event, that the filewatcher (the node function is `fs.watch`) is detecting, it is not delete but `rename` here on my side.
![image](https://user-images.githubusercontent.com/11188741/118079286-33d6c680-b3b8-11eb-874b-eff74ec7f7f0.png)

I'm AFK for two weeks now and will look deeper into this after. If you have proper knowledge on Javascript and node.js you can feel free to fork sleek, run it from source and try to take a look at the file watcher implementation. The file watcher itself is in `main.js` in a constant function named `startFileWatcher`.

**Comment by ransome1** [comment]
@dekonst  I was able to kind of able to reproduce it, though I havn't found a proper solution yet. 

To double check: This issue only occurs after a change is being syncronised from Androids Syncthing to your computer, correct? Or are you experiencing any issues while syncing between computers using Synchting as well?

**Comment by ransome1** [comment]
@dekonst, I replaced the responsible nodejs method (_fs.watch_) with an actively developed node module named _chokidar_. This module seems to be able to detect when the Android/Syncthing combination is breaking/unlinking the file watcher. Like this I can react to this and simply restart the file watcher.

I created a PoC and after bit of testing it looks like it could be the way to go here. But I would like to ask you to help me test the implementation in your environment, where the issue occured initially. Please feel free to download this pre-release https://github.com/ransome1/sleek/releases/tag/v1.0.4-1 and give me some feedback, if it changes anything on your side as well. 

Usually replacing such a critical function will solve one issue and create several new. Let's see where we end up ;)

**Comment by ransome1** [comment]
@dekonst, have you already found some time to check it?

**Comment by dekonst** [comment]
@ransome1, sorry for the delayed answer.

If you'll run Syncthing on your local computer with the debug facilities switched on for the file system events ([Debugging Syncthing — Syncthing v1 documentation](https://docs.syncthing.net/dev/debugging.html#debugging)):

`$ STTRACE=fs syncthing`

then you may see that Syncthing creates new temporary file, sends content of the changed remote file into that temporary file, then removes original file and renames temporary file into original file name:

![image](https://user-images.githubusercontent.com/33258381/120364896-75eb8c00-c316-11eb-837a-1485aafa67ad.png)

So it means that this is common issue not related to the Android and it should happen while syncing between computers using Synchting as well.

I've checked your changes in my environment both on version v1.0.4-1, and v1.0.5. Everything is ok. So, I confirm that the issue is closed.

**Comment by ransome1** [comment]
Glad to here it solves the issue, as this use case is quite important to me.
